src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/PileupSpark.java

```diff
@@ -0,0 +1,152 @@
+                    new ShufflePileupFunction(getHeaderForReads()));
+        } else {
+            alignmentContexts = SparkUtils.joinOverlapping(ctx, reads, GATKRead.class, getBestAvailableSequenceDictionary(),
+                    intervalShards, new PileupFunction(getHeaderForReads()));
+        }
+
+        Broadcast<ReferenceMultiSource> bReferenceSource = hasReference() ? ctx.broadcast(getReference()) : null;
+        alignmentContexts.map(alignmentContext -> {
+            // TODO: this reference lookup is potentially slow, push down to do once per interval shard
+            byte[] bases = hasReference() ? bReferenceSource.getValue().getReferenceBases(null, new SimpleInterval(alignmentContext.getLocation())).getBases() : null;
+            final ReadPileup basePileup = alignmentContext.getBasePileup();
+            return basePileup.getPileupString(bases != null ? (char) bases[0] : 'N');
+        }).saveAsTextFile(outputFile);
+    }
+
+    // similar to ReadShard, except there is no underlying ReadSource
+    private List<SimpleInterval> getIntervalShards(List<SimpleInterval> intervals, int intervalShardSize) {
+        final List<SimpleInterval> shards = new ArrayList<>();
+        for (SimpleInterval interval : intervals) {
+            int start = interval.getStart();
+            while (start <= interval.getEnd()) {
+                int end = Math.min(start + intervalShardSize - 1, interval.getEnd());
+
+                final SimpleInterval nextShardInterval = new SimpleInterval(interval.getContig(), start, end);
+                shards.add(nextShardInterval);
+
+                start += intervalShardSize;
+            }
+        }
+        return shards;
+    }
+
+    @SuppressWarnings("unchecked")
+    private static class PileupFunction implements FlatMapFunction2<Iterator<GATKRead>, Iterator<SimpleInterval>, AlignmentContext> {
+        private static final long serialVersionUID = 1L;
+
+        private SAMFileHeader header;
+
+        public PileupFunction(SAMFileHeader header) {
+            this.header = header;
+        }
+
+        @Override
+        public Iterable<AlignmentContext> call(Iterator<GATKRead> readIterator,
+                                                         Iterator<SimpleInterval> intervalIterator) {
+            // get the samples from the read groups
+            final Set<String> samples = header.getReadGroups().stream()
+                    .map(SAMReadGroupRecord::getSample)
+                    .collect(Collectors.toSet());
+            LocusIteratorByState libs = new LocusIteratorByState(readIterator, DownsamplingMethod.NONE, true, false, false, samples, header);
+            return new IntervalOverlappingIterator<>(libs, ImmutableList.copyOf(intervalIterator), header.getSequenceDictionary());
+        }
+    }
+
+    private static class ShufflePileupFunction implements FlatMapFunction<Tuple2<SimpleInterval, Iterable<GATKRead>>, AlignmentContext> {
```

Comment:
Thanks a lot for the answer @tomwhite!

Actually it would be nice to have a way of using an implemented engine for a `LocusWalker` work with Spark. Thus, I think that this requires to have the reference bases for each context. But I don't know if this is out of the scope of this PR or if it is doable here. I would appreciate a lot if it could be done :)